docs/docs/specs/v0/composer-2.md

```diff
@@ -0,0 +1,389 @@
+- The main `<App />` component also wraps the main `Surface` with an `AppContext` provider representing the total and entire state of the application which becomes available through a hook to any nested component.
+
+## Example Structure
+
+Below is a diagram illustrating how:
+
+- a full application is built up to handle a `Stack` of `Composer` and `Image`
+- `Surfaces` are the only element of abstraction necessary to both compose the entire application chrome and implement the abstract portion of `Stack`.
+
+![Composer Structural Diagram](./composer.drawio.svg)
+
+## Stacks
+
+- Stacks are just components which rely on the same `<Surface />` concept to delegate the concrete rendering of stack elements to components provided by plugins.
+
+## Routing
+
+- Routing (the main router and top level routes) are declared outside (around) the `<App />` component and allow the `App` to pass route information as state in the total `AppState` context for all descendants.
+- Nested components inside `<App />` are equally free to use `<Route />` elements if necessary as well.
+- Nested components are encouraged to perform navigation by submitting `Actions` as opposed to calling `navigate()` directly, although situations like `<Button to="/some/route"/>` should also be supported.
+
+## Application State and Actions
+
+- The App provides a global `AppState` (in the `AppContext`) to all nested components which is composed of some concrete and other dynamic sections to be filled by plugins
+- State is updated via a `dispatch(...actions)` API in the `AppContext` which enables Plugins to subscribe to each other's actions and update state accordingly
+- Plugins (and their provided Components) are free to extend/add more fields to the `AppState` as necessary, when it is desirable to share certain state application-wide.
+- State can be composed of `@preact/signals` (or similar reactive state solution) and can contain any other objects including `ECHO` objects.
+- When state is:
+  - local, ephemeral, per-session: it is managed by signals (and backed by local or session storage as appropriate). i.e. the selection state in a List or Tree is ephemeral and not shared between windows or devices, but may be shared among components and plugins in the application.
+  - peer-to-peer, permanent: it is managed by ECHO objects.
+- `AppState` includes current route and params in convenient form.
+- Actions are handled by functions that resemble `reducers` who should return new state reflecting the action committing, and possibly a list of side-effects which may result in more actions or state updates at a later time. (i.e. performing writes to echo objects in a batch is a side-effect, invoking a github API is a side-effect, etc. whereas the indicator that the app is loading something is a purely new state returned from the reducer)
```

Comment:
A component can have a reducer if it likes but I don’t think we should impose that pattern on strictly all components. If `AppState` is a set of signals, it’s sufficient for a component to subscribe to changes on a certain signal and subsequently dispatch changes to signals as it likes, i.e. all actions would just update signals.

Reply:
consider for example a plugin which prefixes all navigate requests (or rewrites the urls) during action invocation (maybe a bad example) - but you could intercept requests for navigation from say buttons or hyperlinks in deep surfaces, and alter their destinations to be routed to a local version of halo (if the link is pointing to a halo url for example)

Reply:
> i.e. all actions would just update signals.
some actions may not update signals and/or may encode other side effects

Reply:
> for example a plugin which prefixes all navigate requests

Navigation is mutating the route, a substate of `AppState`, so this is still a global state mutation

> may not update signals and/or may encode other side effects

I can see the need for that, but that may be an action that the extensibility engine would not know what to do with and expect a plugin to handle. I suppose what I meant by “all actions would update signals”, I meant “the extensibility engine only responds to actions which update signals”, since any other behavior is probably out of scope for the framework and within-scope for the plugins.